keps/sig-api-machinery/0032-20181026-flexible-inflight-request-limit.md

```diff
@@ -0,0 +1,153 @@
+
+Combination of `limit` and `request` allows to prevent overreservation of resources for critical users on the one hand, and flexible restrict different parts of the system from overspending resources on the other hand.
+
+Using limits in `tokens` instead of limiting number of requests allows to
+account resource consumption more realistic.
+
+
+### Goals <a name="goals"></a>
+
+* Implement per-user and per-group `limit` and `request`
+* Implement request cost estimation
+
+## Proposal <a name="proposal"></a>
+
+Implement weight-based per-user and per-group limits instead of current count-based limits.
+Every request has its own `weight` measured in `tokens`, according to its complexity. Exact algorithm of request cost calculation described below.
+
+Implement options `readonly_capacity`, `mutating_capacity`: Total `weight` of inflight requests that apiserver could carry.
+
+Implement new API resource `InflightLimit` to store group limits and requests:
+```
+apiVersion: extensions/v1beta1
+kind: InflightLimit
+metadata:
+  name: %NAME%
+spec:
+  subjects:
+    kind: %SUBJECT_KIND%
+    name: %SUBJECT_NAME%
+  readonly:
+    limit: %READONLY_TOKENS_LIMIT%
+    request: %READONLY_TOKENS_REQUEST%
+  mutating:
+    limit: %MUTATING_TOKENS_LIMIT%
+    request: %MUTATING_TOKENS_REQUEST%
+status:
+  readonlyInflightTokens: %CURRENT_READONLY_TOKENS_INFLIGHT%
+  mutatingInflightTokens: %CURRENT_MUTATING_TOKENS_INFLIGHT%
+```
+
+Where:
+
+`%NAME%` - The name of `InflightLimit` resource
+
+`%SUBJECT_KIND%` - `User` or `Group`, like in `RoleBinding`
+
+`%SUBJECT_NAME%` - User or group name
+
+`%READONLY_TOKENS_LIMIT%` - Limits total readonly requests `weight` across  particular user or group
+
+`%READONLY_TOKENS_REQUEST%` - Reserved (always available) `weight` of readonly requests  for particular user or group
+
+`%MUTATING_TOKENS_LIMIT%` - Limits total mutating requests `weight` across  particular user or group.
+
+`%MUTATING_TOKENS_REQUEST%` - Reserved (always available) `weight` of mutating requests  for particular user or group
+
+`%CURRENT_READONLY_TOKENS_INFLIGHT%` - Total `weight` of inflight readonly requests from the user or group
+
+`%CURRENT_MUTATING_TOKENS_INFLIGHT%` - Total `weight` of inflight mutating requests from the user or group
+
+**Example**: We have `readonly_capacity` = 1000. `readonly.request` specified for `system:masters` is 50, for user `node` is 100, so total available readonly request `tokens` for other users is 850. In other words, members of `system:masters` in any time have at least 50 `tokens`, user `node` in any time have at least 100 `tokens`.
+
+
+### Algorithm to accept/reject readonly request <a name="algorithm-to-accept-reject-readonly-request"></a>
+
+0. User $CURRENT_USER made readonly request to apiserver, pick the request `weight` as `$REQUEST_WEIGHT`
```

Comment:
My first instinct is something like this:

0. Map user/group/namespace -> quota bucket (details should be covered in a different section)
1. Compute EC = estimated request cost (upper bound) (details should be covered in a different section)
2. Compute T1 = tier 1 quota remaining for this user and T2 = tier 2 quota available in the system for use by this user.
3. Compare EC with T1: `if EC <= T1`, accept request and deduct quota (`T1-=EC`)
4. Compare EC with T1 + T2; `if EC < T1+T2`, accept request & deduct quota (`T1 = 0; T2 -= (EC - T1)`)
4. (optional) If request is accepted, compute actual cost and refund the difference.

Note that EC, T1, and T2 are parameterized by request type, one of: `[readonly, mutating, long-running]`. The quota distribution algorithm should be the same for each unless we have a good reason for it to not be the same.